### DISCUSSION TITLE
请问 quarkd 是怎么做到服务端渲染的

### DISCUSSION THREAD

**Discussion by lishaobos** [hellof2e/quark-design#204] [Q&A]
# 请问 quarkd 是怎么做到服务端渲染

**Comment by yanhaijing** [comment]
做不到吧，我了解到的是webcomponents方案，都没法支持SSR，可以搜索下webcomponents怎么支持ssr

**Reply by BugMaker-Huang** [reply]
错误的，web components方案是有办法支持服务端渲染的；参见：
https://stenciljs.com/docs/hydrate-app
https://github.com/lit/lit/blob/main/packages/lit-html/package.json#L20

**Reply by yanhaijing** [reply]
好的感谢，我研究研究，webcomponents本身的declarative shadow dom 还不太可用吧

**Comment by BugMaker-Huang** [comment]
@yanhaijing 
> 好的感谢，我研究研究，webcomponents本身的declarative shadow dom 还不太可用吧

从源码上看，lit的ssr方案走的就是DSD(Declarative Shadow DOM)

> React server rendering of custom elements will only shallow render components by emitting the host element tag with props stringified to attributes and any light DOM children it has. @lit-labs/ssr is able to server render Lit elements deeply by emitting the contents of the component as Declarative Shadow DOM. We want usage of custom elements in React with server rendering to be able to take advantage of this as well.

https://github.com/lit/lit/blob/main/dev-docs/design/ssr-react.md#background
https://github.com/lit/lit/tree/main/packages/labs/ssr

**Comment by BugMaker-Huang** [comment]
不过即使不引入DSD，只要在打包构建的流程上，保证不使用'windows'变量，替换成globalThis；然后适当加入[一些补丁](https://github.com/lit/lit/tree/main/packages/labs/ssr-dom-shim)，就可以解决像[这样的问题](https://github.com/hellof2e/quark-design/issues/215)。

只不过这么做的话当然会导致hydration完成前，组件只有light-dom；但至少项目能跑起来了。

**Reply by yanhaijing** [reply]
嗯嗯，大概思路就是把dom api 替换下，ssr里面改成输出字符串
